graphql.go

```diff
@@ -0,0 +1,148 @@
+	RequestedQueryCost int                   `json:"requestedQueryCost"`
+	ActualQueryCost    *int                  `json:"actualQueryCost"`
+	ThrottleStatus     GraphQLThrottleStatus `json:"throttleStatus"`
+}
+
+// GraphQLThrottleStatus represents the status of the shop's rate limit points
+type GraphQLThrottleStatus struct {
+	MaximumAvailable   float64 `json:"maximumAvailable"`
+	CurrentlyAvailable float64 `json:"currentlyAvailable"`
+	RestoreRate        float64 `json:"restoreRate"`
+}
+
+type graphQLError struct {
+	Message    string                 `json:"message"`
+	Extensions *graphQLErrorExtension `json:"extensions"`
+	Locations  []graphQLErrorLocation `json:"locations"`
+}
+
+type graphQLErrorExtension struct {
+	Code          string
+	Documentation string
+}
+
+const (
+	graphQLErrorCodeThrottled = "THROTTLED"
+)
+
+type graphQLErrorLocation struct {
+	Line   int `json:"line"`
+	Column int `json:"column"`
+}
+
+// Query creates a graphql query against the Shopify API
+// the "data" portion of the response is unmarshalled into resp
+func (s *GraphQLServiceOp) Query(q string, vars, resp interface{}) error {
+	data := struct {
+		Query     string      `json:"query"`
+		Variables interface{} `json:"variables"`
+	}{
+		Query:     q,
+		Variables: vars,
+	}
+
+	attempts := 0
+
+	for {
+		gr := graphQLResponse{
+			Data: resp,
+		}
+
+		err := s.client.Post("graphql.json", data, &gr)
+		// internal attempts count towards outer total
+		attempts += s.client.attempts
+		s.client.attempts = attempts
+
+		var ra float64
+
+		if gr.Extensions != nil {
+			ra = gr.Extensions.Cost.RetryAfterSeconds()
+			s.client.RateLimits.GraphQLCost = &gr.Extensions.Cost
+			s.client.RateLimits.RetryAfterSeconds = ra
+		}
+
+		if len(gr.Errors) > 0 {
+			re := ResponseError{Status: 200}
+			var doRetry bool
+
+			for _, err := range gr.Errors {
+				if err.Extensions != nil && err.Extensions.Code == graphQLErrorCodeThrottled {
+					if attempts >= s.client.retries {
+						return RateLimitError{
+							RetryAfter: int(ra),
+							ResponseError: ResponseError{
+								Status:  200,
+								Message: err.Message,
+							},
+						}
+					}
+
+					// only need to retry graphql throttled retries
+					doRetry = true
+				}
+
+				re.Errors = append(re.Errors, err.Message)
+			}
+
+			if doRetry {
+				wait := time.Duration(ra) * time.Second
+				s.client.log.Debugf("rate limited waiting %s", wait.String())
+				time.Sleep(wait)
+				continue
+			}
+
+			err = re
+		}
+
+		return err
+	}
+}
+
+// RetryAfterSeconds returns the estimated retry after seconds based on
+// the requested query cost and throttle status
+func (c GraphQLCost) RetryAfterSeconds() float64 {
+	var diff float64
+
+	if c.ActualQueryCost != nil {
+		diff = c.ThrottleStatus.CurrentlyAvailable - float64(*c.ActualQueryCost)
+	} else {
+		diff = c.ThrottleStatus.CurrentlyAvailable - float64(c.RequestedQueryCost)
+	}
+
+	if diff < 0 {
+		return -diff / c.ThrottleStatus.RestoreRate
```

Comment:
This is interesting.

For instance, if you have `CurrentlyAvailable == 5` and `ActualQueryCost == 110` then the diff is 105 and you need 3 seconds to replish more than 105 (3 * 50)

From the docs:
`This quota replenishes at a rate of 50 cost points per second.`

Do we know that waiting for fractions of a second replenishes a fraction of 50 cost points? 
If not then you might as well always round up to the next full second and make sure you wait long enough to accumulate the points needed for the next request 

On the other hand, if you make the request too early and you'll make ore http requests but possibly make the "right" request a bit earlier.
Interesting trade-off.

Thoughts?

Reply:
I believe that waiting for 0.5 seconds won't give us a 25 cost point replenishment

I updated the wait to use the ceiling of the calculated wait time, so that we're guaranteed to have replenished enough

Reply:
👍